File: solutions/ocaml/binary-search-tree/1/binary_search_tree.ml
(* Dedicated to Junko F. Didi and Shree DR.MDD *)

open Base

type bst = Nil | Node of (int * bst * bst)

let empty = Nil

let value = function
  | Nil -> Result.fail "empty"
  | Node (num, _, _) -> Result.return num

let left = function
  | Nil -> Result.fail "empty"
  | Node (_, lnode, _) -> Result.return lnode

let right = function
  | Nil -> Result.fail "empty"
  | Node (_, _, rnode) -> Result.return rnode

let rec insert elem = function
  | Nil -> Node (elem, Nil, Nil)
  | Node (pivot, lnode, rnode) ->
      if elem <= pivot
      then Node (pivot, insert elem lnode, rnode)
      else Node (pivot, lnode, insert elem rnode)

let rec to_list = function
  | Nil -> []
  | Node (pivot, lnode, rnode) ->
      to_list lnode @ (pivot :: to_list rnode)

File: solutions/fsharp/binary-search-tree/1/BinarySearchTree.fs
// For my Shree DR.MDD
module BinarySearchTree

type Node = { left: Node option; value: int; right: Node option }

let left n  = n.left
let right n = n.right
let data n = n.value

let private makeNode v = { left = None; right = None; value = v }

let rec private insert v (tree: Node) =
    let attach valToAdd branch =
        match branch with
        | Some child -> Some (insert valToAdd child)
        | None -> Some (makeNode valToAdd)

    if v <= tree.value then
        { tree with left = attach v tree.left }
    else
        { tree with right = attach v tree.right }

let sortedData tree = 
    let rec traverse =
        function
        | Some n -> traverse n.left @ [n.value] @ traverse n.right
        | None -> []
    traverse (Some tree)

let create = 
    function
    | [] -> failwith "Cannot create tree from empty list."
    | h::t -> List.fold (fun acc x -> insert x acc) (makeNode h) t
